src/librustc_ast_passes/ast_validation.rs

```diff
@@ -883,30 +882,54 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if is_auto == IsAuto::Yes {
                     // Auto traits cannot have generics, super traits nor contain items.
                     if !generics.params.is_empty() {
-                        struct_span_err!(
+                        let mut err = struct_span_err!(
                             self.session,
-                            item.span,
+                            generics.span,
                             E0567,
                             "auto traits cannot have generic parameters"
-                        )
-                        .emit();
+                        );
+                        err.span_label(
+                            item.ident.span,
+                            "auto trait cannot have generic parameters",
+                        );
+                        err.span_suggestion(
+                            generics.span,
+                            "remove the parameters",
+                            String::new(),
+                            Applicability::MachineApplicable,
+                        );
+                        err.emit();
```

Comment:
Some points:
1. This is now chainable, so let's.
2. I think this is sufficiently long to deserve it's own method (also to make the visiting logic easier to read).